src/workerd/io/worker.c++

```diff
@@ -1145,6 +1145,96 @@ Worker::Isolate::Isolate(kj::Own<Api> apiParam,
+    // resolved or rejected, we check the tag also. If the promise tag and the current
+    // isolate tag do not match, the function below is called.
+    if (features.getHandleCrossRequestPromiseResolution()) {
+      lock->v8Isolate->SetPromiseCrossContextResolveCallback(
+          [](v8::Isolate* isolate, v8::Local<v8::Value> tag, v8::Local<v8::Data> reactions,
+              v8::Local<v8::Value> argument,
+              std::function<void(v8::Isolate * isolate, v8::Local<v8::Data> reactions,
+                  v8::Local<v8::Value> argument)> callback) {
+        try {
+          auto& js = jsg::Lock::from(isolate);
+          // The promise tag is generally opaque except for right here. The tag
+          // wraps an instanceof kj::Own<IoCrossContextExecutor>, which wraps an atomically
+          // refcounted pointer to the DeleteQueue for the correct isolate.
+          // We simply pass the given callback, reactions, and argument to
+          // a function that will be added to the queue inside DeleteQueue.
+          // The next time the relevant IoContext is entered, this queue will
+          // be drained and the actions will be run. Adding the task to the
+          // delete queue will also signal the IoContext that it should wake
+          // up and drain the queue. Simple, eh?
+          //
+          // A word of warning tho! It is possible for the IoContext to be
+          // destroyed before the promise is resolved. Any actions that have
+          // already been added to the queue would end up being dropped silently
+          // on the floor. Actions that are added to the queue now will be run
+          // immediately in the wrong IoContext.
+          auto& ref = jsg::unwrapOpaqueRef<kj::Own<IoCrossContextExecutor>>(isolate, tag);
+          ref->execute(js,
+              [reactions = v8::Global<v8::Data>(isolate, reactions),
+                  argument = jsg::JsRef(js, jsg::JsValue(argument)),
+                  callback = kj::mv(callback)](jsg::Lock& js) mutable {
+            try {
+              callback(js.v8Isolate, reactions.Get(js.v8Isolate), argument.getHandle(js));
+            } catch (jsg::JsExceptionThrown&) {
+              // Exceptions here are generally unexpected but possible. Let's just
+              // return control back to V8 which should handle the case.
+            } catch (...) {
+              auto ex = kj::getCaughtExceptionAsKj();
+              KJ_LOG(ERROR, "Scheduling deferred promise continuations failed unexpectedly", ex);
+              jsg::throwInternalError(js.v8Isolate, kj::mv(ex));
+            }
+          });
+        } catch (jsg::JsExceptionThrown&) {
+          // Exceptions here are generally unexpected but possible because the jsg::Promise
+          // then can fail if the isolate is in the process of being torn down. Let's just
+          // return control back to V8 which should handle the case.
```

Comment:
In this case we are in fact returning to V8. But will V8 actually do the right thing if an exception is scheduled during the callback? Normally in V8 code we would return an empty handle or empty maybe to signal that an exception is scheduled, so that the caller can return early. Maybe the callback should return a boolean so that the call site can do the right thing in the exception case?

Reply:
Got it figured out I think. V8 was not properly checking if an exception was being scheduled in the reject path causing it to not properly handle errors here. I *think* the patch now covers the relevant case but it's possible there are other places? I need to keep digging but let's see if we can make progress on landing this now